metaflow/plugins/tagging_cli.py

```diff
@@ -0,0 +1,457 @@
+        )
+
+    flow_name = obj.flow.name
+
+    # handle error messaging for two cases
+    # 1. our user tries to tagging a new flow before it is run
+    # 2. our user makes a typo in --namespace
+    try:
+        namespace(user_namespace)
+        Flow(pathspec=flow_name)
+    except MetaflowNotFound:
+        raise CommandException(
+            "No execution found for *%s*. Please run the flow before tagging."
+            % flow_name
+        )
+
+    except MetaflowNamespaceMismatch:
+        raise CommandException(
+            "No execution found for *%s* in namespace *%s*."
+            % (flow_name, user_namespace)
+        )
+
+    # throw an error with message to include latest run-id when run_id is None
+    if run_id is None:
+        latest_run_id = Flow(pathspec=flow_name).latest_run.id
+        msg = (
+            "Please specify a run-id using --run-id.\n"
+            "*%s*'s latest execution in namespace *%s* has id *%s*."
+            % (flow_name, user_namespace, latest_run_id)
+        )
+        raise CommandException(msg)
+    run_id_parts = run_id.split("/")
+    if len(run_id_parts) == 1:
+        run_name = "%s/%s" % (flow_name, run_id)
+    else:
+        raise CommandException("Run-id *%s* is not a valid run-id" % run_id)
+
+    # handle error messaging for three cases
+    # 1. our user makes a typo in --run-id
+    # 2. our user tries to mutate tags on a meson/production run
+    # 3. our user's --run-id does not exist in the default/specified namespace
+    try:
+        namespace(user_namespace)
+        metaflow_client = Run(pathspec=run_name)
+    except MetaflowNotFound:
+        raise CommandException("No run *%s* found" % run_name)
+    except MetaflowNamespaceMismatch:
+        namespace(None)
+        metaflow_client = Run(pathspec=run_name)
+
+        run_sys_tags = metaflow_client.system_tags
+
+        if any([is_prod_token(tag) for tag in run_sys_tags]):
+            raise CommandException(
+                "Modifying a scheduled run's tags is currently not allowed."
+            )
+
+        msg = "Run *%s* does not belong to namespace *%s*\n" % (
+            run_name,
+            user_namespace,
+        )
+
+        user_tags = [tag for tag in run_sys_tags if tag.startswith("user:")]
+        if user_tags:
+            msg += (
+                "You can choose any of the following tags associated with *%s* "
+                "to use with option *--namespace*\n" % run_name
+            )
+            msg += "\n".join(["\t*%s*" % tag for tag in user_tags])
+
+        raise CommandException(msg)
+
+    return metaflow_client
+
+
+def list_runs(obj, user_namespace):
+    namespace(user_namespace)
+    metaflow_client = Flow(pathspec=obj.flow.name)
+    return metaflow_client.runs()
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.group(help="Commands related to Metaflow tagging.")
+def tag():
+    pass
+
+
+@tag.command("add", help="Add tags to a run.")
+@click.option(
+    "--run-id",
```

Comment:
How about simplifying the logic in `get_run_client` by making one of `--latest-successful-run`, `--latest-run` and `--run-id` required?